File: défi mensuel/2025/03/soluce.rb
class Matrice

    #attr_accessor :ordre, :list, :max

    def initialize(ordre = 100)
        @ordre = ordre
        @list  = Array.new(ordre) { Array.new(ordre, 0) }
        @max   = 0
    end

    def to_s
        txt = 'Matrice'
        maxLen = @max.to_s.length
        @ordre.times do |j|
            txt += "\n"
            @ordre.times do |i|
                item = @list[j][i]
                iLen = item.to_s.length
                if   iLen < maxLen then txt += " "*(maxLen - iLen) end
                txt += item.to_s + " "
                end
            end
        return txt
    end

    def verifierIndex(i, j)
        if i >= @ordre then raise IndexError.new("indice [i] au-delà des limites") end
        if j >= @ordre then raise IndexError.new("indice [j] au-delà des limites") end
        if i < 0 then return self.verifierIndex(@ordre + i, j) end
        if j < 0 then return self.verifierIndex(i, @ordre + j) end
        return [i, j]
    end

    def [](i, j)
        bi, bj = self.verifierIndex(i, j)
        return @list[bj][bi]
    end

    def incrementer(i, j)
        bi, bj = self.verifierIndex(i, j)
        @list[bj][bi] += 1
        if @list[bj][bi] > @max then @max = @list[bj][bi] end
    end

    def appartientCercle(i, j)
        bi, bj   = self.verifierIndex(i, j)
        centre   = (@ordre - 1) / 2
        distance = (centre - bi)**2 + (centre - bj)**2
        rayon    = (@ordre / 2)**2
        return distance <= rayon
    end

    def getOrdre
        return @ordre
    end

    def getMax
        return @max
    end

    def each
        j = 0
        @ordre.times do
            i = 0
            @ordre.times do
                yield [@list[j][i], i, j]
                i += 1
            end
            j += 1
        end
    end
end

def approcherPI(matrice)
    if matrice.getMax == 0 then raise StandardError.new("Pas de lancers dans la matrice.") end
    countOUT, countIN = [0.0, 0.0]
    matrice.each { |item, i, j| if matrice.appartientCercle(i, j) then countIN += item.to_f else countOUT += item.to_f end }
    return ((4 * countIN) / (countOUT + countIN))
end

def simulerLancer(matrice, n = 200)
    b = matrice.getOrdre - 1
    n.times do matrice.incrementer(rand(0..b), rand(0..b)) end
end

A = Matrice.new(10)
#puts A
simulerLancer(A, 20000)
#puts A
puts approcherPI(A)

File: défi mensuel/2025/03/soluce.py
class Matrice :
    """Matrice simulée avec une liste en 2D."""
    
    def __init__(self, ordre : int = 100):
        self.ordre : int  = ordre
        self.list  : list = [ [ 0 for _ in range(0, ordre) ] for _ in range(0, ordre) ]
        self.max   : int = 0
        
    def vérifierIndex(self, i : int, j : int) -> tuple :
        if i >= self.ordre or j >= self.ordre : raise IndexError("soit i soit j en dehors de la plage des indices valables")
        if i < 0 : i = self.ordre + i
        if j < 0 : j = self.ordre + j
        return(i, j)        

    def __getitem__(self, i : int, j : int) -> int :
        i, j = self.vérifierIndex(i, j)
        return self.list[j][i]
    
    def incrémenter(self, i : int, j : int):
        i, j = self.vérifierIndex(i, j)
        self.list[j][i] += 1
        if self.list[j][i] > self.max : self.max = self.list[j][i]
        
    def appartientCercle(self, i : int, j : int) -> bool :
        i, j = self.vérifierIndex(i, j)
        centre   : float = (self.ordre - 1) / 2
        distance : float = (centre - i)**2 + (centre - j)**2
        rayon    : float = (self.ordre / 2)**2
        return distance <= rayon
    
    def __str__(self) -> str :
        txt : str = "Matrice"
        maxLen : int = len(str(self.max))
        for j in self.list :
            txt += '\n'
            for i in j     :
                iLen : int = len(str(i))
                if iLen < maxLen :
                       txt += " "*(maxLen-iLen) + str(i)
                else : txt += str(i)
                txt += ' '
        return txt
    
    def getOrdre(self) -> int : return self.ordre
    
    def __iter__(self) :
        for j in range(0, self.ordre):
            for i in range(0, self.ordre):
                yield (i, j, self.list[j][i])

def approcherPI(matrice : 'Matrice') -> float :
    """À partir d'une matrice, approche π."""
    if matrice.max == 0 : raise ValueError("Pas de lancers dans la matrice.")
    count   : int = 0
    countIn : int = 0
    for element in matrice :
        pos : bool = matrice.appartientCercle(element[0], element[1])
        if pos : countIn += element[2]
        count += 1
    return 4 * (countIn / count)

def simulerLancer(matrice : 'Matrice', n : int = 1):
    """Simule n lancers aléatoires dans la matrice donnée."""
    from random import randint
    for _ in range(n): matrice.incrémenter(randint(0, matrice.getOrdre()-1), randint(0, matrice.getOrdre()-1))

# Création de la matrice
A = Matrice(100)
# Nombre de lancers dans la matrice - essayez avec une valeur de n différente
simulerLancer(A, 20000)
# Approche de π
print("π ≈ ", approcherPI(A), "; π = 3.14159")
